### ISSUE TITLE
assigning_clones: `clone_into()` would cause lifetime error

### ISSUE THREAD

**Issue by anatawa12** [rust-lang/rust-clippy#13070]
### Summary

code like `bin_name = bin_name[slash + 1..].to_owned();` would trigger `assigning_clones` warning but fixing it would cause lifetime error.

### Lint Name

assigning_clones

### Reproducer

I tried this code:

```rust
fn lint_err(mut bin_name: String) {
    if let Some(slash) = bin_name.rfind(&['/', '\\']) {
        bin_name = bin_name[slash + 1..].to_owned();
    }
}
```

I saw this happen:

```

warning: assigning the result of `ToOwned::to_owned()` may be inefficient
    --> vrc-get/src/commands.rs:1449:9
     |
1449 |         bin_name = bin_name[slash + 1..].to_owned();
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `clone_into()`: `bin_name[slash + 1..].clone_into(&mut bin_name)`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assigning_clones
     = note: `#[warn(clippy::assigning_clones)]` on by default
```

I expected to see this happen:

No lint warning since `bin_name[slash + 1..].clone_into(&mut bin_name)` would cause lifetime error

### Version

```text
rustc 1.79.0 (129f3b996 2024-06-10)
binary: rustc
commit-hash: 129f3b9964af4d4a709d1383930ade12dfe7c081
commit-date: 2024-06-10
host: aarch64-apple-darwin
release: 1.79.0
LLVM version: 18.1.7
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by y21** [comment]
This was fixed in #12756: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=62a13bf97bcc4bde6b34bb3401abeede)

It'll take a bit for the fix to land on stable however (it's currently only on nightly, in version 1.81)

**Comment by anatawa12** [comment]
Sorry fpr duplicate. I searched with "assigning_clones lifetime" but other issues used borrow checker so I missed previous ones.
